File: trunk/ServerPhp/src/dc/DC0064.php
<?php
/* N21 eBusiness Suite
 * Copyright: Nan21 Electronics srl
 * Generated content.
 * DC0064 Controller: Purchase order lines
 */

require_once(PATH_N21LIB."/Http/Controller.php");

class DC0064 extends Controller {

  public function __construct($request, $response, $session) {
    parent::__construct($request, $response, $session);
    $this->_initFields();
  }

private function preQuery(&$params, &$where) {
    if (!empty($_REQUEST["QRY_CURRENCY_CODE"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "pol.CURRENCY_CODE like :CURRENCY_CODE";
      $params["CURRENCY_CODE"] = $_REQUEST["QRY_CURRENCY_CODE"];
      $params["CURRENCY_CODE"] = strtoupper($params["CURRENCY_CODE"]);
    }
    if (!empty($_REQUEST["QRY_DATE_DELIVERED"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "pol.DATE_DELIVERED like :DATE_DELIVERED";
      $params["DATE_DELIVERED"] = $_REQUEST["QRY_DATE_DELIVERED"];
    }
    if (!empty($_REQUEST["QRY_DATE_PROMISED"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "pol.DATE_PROMISED like :DATE_PROMISED";
      $params["DATE_PROMISED"] = $_REQUEST["QRY_DATE_PROMISED"];
    }
    if (!empty($_REQUEST["QRY_DATE_REQUESTED"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "pol.DATE_REQUESTED like :DATE_REQUESTED";
      $params["DATE_REQUESTED"] = $_REQUEST["QRY_DATE_REQUESTED"];
    }
    if (!empty($_REQUEST["QRY_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "pol.ID like :ID";
      $params["ID"] = $_REQUEST["QRY_ID"];
    }
    if (!empty($_REQUEST["QRY_LINE_NR"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "pol.LINE_NR like :LINE_NR";
      $params["LINE_NR"] = $_REQUEST["QRY_LINE_NR"];
    }
    if (!empty($_REQUEST["QRY_NET_AMOUNT"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "pol.NET_AMOUNT like :NET_AMOUNT";
      $params["NET_AMOUNT"] = $_REQUEST["QRY_NET_AMOUNT"];
    }
    if (!empty($_REQUEST["QRY_NOTES"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "pol.NOTES like :NOTES";
      $params["NOTES"] = $_REQUEST["QRY_NOTES"];
    }
    if (!empty($_REQUEST["QRY_PORDER_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "pol.PORDER_ID like :PORDER_ID";
      $params["PORDER_ID"] = $_REQUEST["QRY_PORDER_ID"];
    }
    if (!empty($_REQUEST["QRY_PRICE"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "pol.PRICE like :PRICE";
      $params["PRICE"] = $_REQUEST["QRY_PRICE"];
    }
    if (!empty($_REQUEST["QRY_PRODUCT_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "pol.PRODUCT_ID like :PRODUCT_ID";
      $params["PRODUCT_ID"] = $_REQUEST["QRY_PRODUCT_ID"];
    }
    if (!empty($_REQUEST["QRY_QTY"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "pol.QTY like :QTY";
      $params["QTY"] = $_REQUEST["QRY_QTY"];
    }
    if (!empty($_REQUEST["QRY_UOM_CODE"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "pol.UOM_CODE like :UOM_CODE";
      $params["UOM_CODE"] = $_REQUEST["QRY_UOM_CODE"];
      $params["UOM_CODE"] = strtoupper($params["UOM_CODE"]);
    }
}

public function doQuery() {
    $start = nvl($this->getRequestParam("start"), 0);
    $limit = nvl($this->getRequestParam("limit"),20);
    $orderBy = (!empty($_REQUEST["sort"]))?$_REQUEST["sort"]:"";
    $orderSense = (!empty($_REQUEST["dir"]))?$_REQUEST["dir"]:"";
    $orderByClause = (!empty($orderBy))? "order by $orderBy $orderSense " : "" ;
    $where = "";
    $params = array();
    $this->preQuery($params, $where);
    if (!empty($where)) {
      $where = " where ".$where;
    }
    $sql = "select 
                pol.CURRENCY_CODE
                ,pol.DATE_DELIVERED
                ,pol.DATE_PROMISED
                ,pol.DATE_REQUESTED
                ,pol.ID
                ,pol.LINE_NR
                ,pol.NET_AMOUNT
                ,pol.NOTES
                ,pol.PORDER_ID
                ,pol.PRICE
                ,pol.PRODUCT_ID
                ,pbo_product.get_name_by_id(pol.product_id,'N') PRODUCT_NAME
                ,pol.QTY
                ,pol.UOM_CODE
            from PURCHASE_ORDER_LINE pol $where $orderByClause ";
    $this->logger->debug($sql);
    $rs = $this->getDbConn()->SelectLimit($sql, $limit, $start, $params);
    $rsCount = $this->getDbConn()->Execute("select count(*) TOTALCOUNT from (".$sql.") t", $params);
    $rsCount->MoveFirst();
    $dataOut = $this->serializeCursor($rs,array_keys($this->fields), $this->getDataFormatFetch());
    if ($this->getDataFormatFetch() == "xml" ) {header("Content-type: application/xml");}
    print $this->formatQueryResponseData($dataOut, $rsCount->fields["TOTALCOUNT"]);
} /* end function doQuery  */


public function doExport() {
    $start = nvl($this->getRequestParam("start"), 0);
    $limit = nvl($this->getRequestParam("limit"),20);
    $groupBy = (!empty($_REQUEST["groupBy"]))?$_REQUEST["groupBy"]:"";
    $orderBy = (!empty($_REQUEST["sort"]))?$_REQUEST["sort"]:"";
    $orderSense = (!empty($_REQUEST["dir"]))?$_REQUEST["dir"]:"";
    $orderByClause = (!empty($orderBy))? "order by $orderBy $orderSense " : "" ;
    $where = "";
    $params = array();
    $this->preQuery($params, $where);
    if (!empty($where)) {
      $where = " where ".$where;
    }
    $sql = "select 
                pol.ID
                ,pol.PORDER_ID
                ,pol.LINE_NR
                ,pbo_product.get_name_by_id(pol.product_id,'N') PRODUCT_NAME
                ,pol.PRODUCT_ID
                ,pol.QTY
                ,pol.UOM_CODE
                ,pol.PRICE
                ,pol.CURRENCY_CODE
                ,pol.NET_AMOUNT
                ,pol.NOTES
                ,pol.DATE_REQUESTED
                ,pol.DATE_PROMISED
                ,pol.DATE_DELIVERED
            from PURCHASE_ORDER_LINE pol $where $orderByClause ";
    $rs = $this->getDbConn()->Execute($sql, $params);
    $rsCount = $this->getDbConn()->Execute("select count(*) TOTALCOUNT from (".$sql.") t", $params);
    $rsCount->MoveFirst();
    if (!empty($_REQUEST["_p_disp_cols"])) {
      $columns = explode("|",$_REQUEST["_p_disp_cols"]);
    } else {
      $columns = array_keys($this->fields);
    }
    if ($this->getDataFormatExport() == "csv" ) {
      $dataOut = $this->serializeCursor($rs,$columns,"csv");
    } else {
      $dataOut = $this->serializeCursor($rs,$columns,"xml");
      $dataOut = "<records>".$dataOut."</records>";
      $dataOut = "<queryParams>".$this->serializeMap($params,"xml")."</queryParams>".$dataOut;
      $dataOut = "<columnDef>".$this->columnDefForExport($columns,$this->fields,true).$this->columnDefForExport(array_diff(array_keys($params), $columns),$this->fields,false)."</columnDef>".$dataOut;
      $dataOut = "<staticText>".$this->exportLocalizedStaticText()."</staticText>".$dataOut;
      $dataOut = "<groupBy>".$groupBy."</groupBy>".$dataOut;
      $dataOut = "<reportData  title=\"".$this->getDcTitle()."\" by=\"".$_SESSION["user"]["userName"]."\" on=\"".date(DATE_FORMAT)."\">".$dataOut."</reportData>";
    }
    $this->beginExport();
    print $dataOut;
    $this->endExport();
} /* end function doExport  */

public function fetchRecord() {
    $this->record["ID"] = $this->getRequestParam("ID");
    if (empty($this->record["ID"])) { throw new Exception("Missing value for primary key field ID in DC0064.fetchRecord().");}
    $pkArray = array("ID" => $this->record["ID"]);
    $this->findByPk($pkArray, $this->record);
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function fetchRecord */

public function doInsert() {
    $this->populateRecordFromRequest(); 
    $sql = "insert into PURCHASE_ORDER_LINE(
                 CURRENCY_CODE
                ,DATE_DELIVERED
                ,DATE_PROMISED
                ,DATE_REQUESTED
                ,ID
                ,LINE_NR
                ,NET_AMOUNT
                ,NOTES
                ,PORDER_ID
                ,PRICE
                ,PRODUCT_ID
                ,QTY
                ,UOM_CODE
            ) values ( 
                 :CURRENCY_CODE
                ,:DATE_DELIVERED
                ,:DATE_PROMISED
                ,:DATE_REQUESTED
                ,:ID
                ,:LINE_NR
                ,:NET_AMOUNT
                ,:NOTES
                ,:PORDER_ID
                ,:PRICE
                ,:PRODUCT_ID
                ,:QTY
                ,:UOM_CODE
    )";
    $stmt = $this->getDbConn()->prepare($sql);
    $this->getDbConn()->Execute($stmt, $this->record);
    $pkArray = array("ID" => $this->record["ID"]);
    $this->findByPk($pkArray, $this->record);
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function doInsert */

public function doUpdate() {
    $this->populateRecordFromRequest();
    if (empty($this->record["ID"])) { throw new Exception("Missing value for primary key field ID in DC0064.doUpdate().");}
    $sql = "update PURCHASE_ORDER_LINE set 
                 ID=:ID
                ,PORDER_ID=:PORDER_ID
                ,LINE_NR=:LINE_NR
                ,PRODUCT_ID=:PRODUCT_ID
                ,QTY=:QTY
                ,UOM_CODE=:UOM_CODE
                ,PRICE=:PRICE
                ,NET_AMOUNT=:NET_AMOUNT
                ,CURRENCY_CODE=:CURRENCY_CODE
                ,NOTES=:NOTES
                ,DATE_REQUESTED=:DATE_REQUESTED
                ,DATE_PROMISED=:DATE_PROMISED
                ,DATE_DELIVERED=:DATE_DELIVERED
    where 
           ID= :ID
    ";
    $stmt = $this->getDbConn()->prepare($sql);
    $this->getDbConn()->Execute($stmt, $this->record);
    $pkArray = array("ID" => $this->record["ID"],"ID" => $this->record["ID"]);
    $this->findByPk($pkArray, $this->record);
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function doUpdate */

public function doDelete() {
    $this->record["ID"] = $this->getRequestParam("ID");
    if (empty($this->record["ID"])) { throw new Exception("Missing value for primary key field ID in DC0064.doDelete().");}
    $sql = "delete from PURCHASE_ORDER_LINE where 
           ID= :ID
    ";
    $stmt = $this->getDbConn()->prepare($sql);
    $this->getDbConn()->Execute($stmt, $this->record);
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function doDelete */

public function initNewRecord() {
    $this->populateRecordFromRequest();
    $this->record["_p_record_status"] = "insert";
    $this->setFieldInitialValues($this->record, "DC0064");
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function initNewRecord */

private function findByPk(&$pkCols, &$record) {
    $sql = "select 
                pol.CURRENCY_CODE
                ,pol.DATE_DELIVERED
                ,pol.DATE_PROMISED
                ,pol.DATE_REQUESTED
                ,pol.ID
                ,pol.LINE_NR
                ,pol.NET_AMOUNT
                ,pol.NOTES
                ,pol.PORDER_ID
                ,pol.PRICE
                ,pol.PRODUCT_ID
                ,pbo_product.get_name_by_id(pol.product_id,'N') PRODUCT_NAME
                ,pol.QTY
                ,pol.UOM_CODE
            from PURCHASE_ORDER_LINE pol
         where 
           pol.ID= :ID
            ";
    $rs = $this->getDbConn()->Execute($sql, $pkCols);
    $row = $rs->FetchRow();
    CollectionUtils::copyMapValues($record, $row);
} /* end function findByPk  */


public function callProcedure($pName) {
  $this->logger->debug("Start: ".$this->dcName.".callProcedure(".$pName.")");
    $this->populateRecordFromRequest();
    print "{success:true, data:".json_encode($this->record)."}";
    $this->logger->debug("End: ".$this->dcName.".callProcedure");
} /* end function callProcedure */


private function _initFields() {
  $this->fields = array(
  "CREATEDBY" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"CREATEDON" => array(parent::FLDPROP_DATA_TYPE => "DATE")
  ,"CURRENCY_CODE" => array(parent::FLDPROP_DATA_TYPE => "STRING",parent::FLDPROP_CASE_RESTRICTION => "Upper")
  ,"DATE_DELIVERED" => array(parent::FLDPROP_DATA_TYPE => "DATE")
  ,"DATE_PROMISED" => array(parent::FLDPROP_DATA_TYPE => "DATE")
  ,"DATE_REQUESTED" => array(parent::FLDPROP_DATA_TYPE => "DATE")
  ,"ID" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"LINE_NR" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"MODIFIEDBY" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"MODIFIEDON" => array(parent::FLDPROP_DATA_TYPE => "DATE")
  ,"NET_AMOUNT" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"NOTES" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"PORDER_ID" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"PRICE" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"PRODUCT_ID" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"PRODUCT_NAME" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"QTY" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"UOM_CODE" => array(parent::FLDPROP_DATA_TYPE => "STRING",parent::FLDPROP_CASE_RESTRICTION => "Upper")
);
}

}
?>

File: trunk/ServerJava/src/net/nan21/ebs/dc/DC0064.java
/* N21 eBusiness Suite
 * Copyright: Nan21 Electronics srl
 * Generated content.
 * DC0064 DC Controller: Purchase order lines
 */

package net.nan21.ebs.dc;


import java.util.*;
import javax.servlet.http.HttpServletResponse;
import net.nan21.lib.*;

public class DC0064 extends AbstractDataControl implements IDataControl {

  public void init (HttpRequest request, HttpServletResponse response, HttpSession session, DbManager dbm) throws Exception {
    this._initFields();
    super.init(request, response, session, dbm);
  }

private void preQuery() {
    if (this.request.getParam("QRY_ID") != null && !this.request.getParam("QRY_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("pol.ID like :ID");
      this.queryParams.put("ID",(String)this.request.getParam("QRY_ID"));
    }
    if (this.request.getParam("QRY_PORDER_ID") != null && !this.request.getParam("QRY_PORDER_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("pol.PORDER_ID like :PORDER_ID");
      this.queryParams.put("PORDER_ID",(String)this.request.getParam("QRY_PORDER_ID"));
    }
    if (this.request.getParam("QRY_LINE_NR") != null && !this.request.getParam("QRY_LINE_NR").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("pol.LINE_NR like :LINE_NR");
      this.queryParams.put("LINE_NR",(String)this.request.getParam("QRY_LINE_NR"));
    }
    if (this.request.getParam("QRY_PRODUCT_ID") != null && !this.request.getParam("QRY_PRODUCT_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("pol.PRODUCT_ID like :PRODUCT_ID");
      this.queryParams.put("PRODUCT_ID",(String)this.request.getParam("QRY_PRODUCT_ID"));
    }
    if (this.request.getParam("QRY_QTY") != null && !this.request.getParam("QRY_QTY").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("pol.QTY like :QTY");
      this.queryParams.put("QTY",(String)this.request.getParam("QRY_QTY"));
    }
    if (this.request.getParam("QRY_UOM_CODE") != null && !this.request.getParam("QRY_UOM_CODE").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("pol.UOM_CODE like :UOM_CODE");
      this.queryParams.put("UOM_CODE",(String)this.request.getParam("QRY_UOM_CODE").toUpperCase());
    }
    if (this.request.getParam("QRY_PRICE") != null && !this.request.getParam("QRY_PRICE").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("pol.PRICE like :PRICE");
      this.queryParams.put("PRICE",(String)this.request.getParam("QRY_PRICE"));
    }
    if (this.request.getParam("QRY_NET_AMOUNT") != null && !this.request.getParam("QRY_NET_AMOUNT").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("pol.NET_AMOUNT like :NET_AMOUNT");
      this.queryParams.put("NET_AMOUNT",(String)this.request.getParam("QRY_NET_AMOUNT"));
    }
    if (this.request.getParam("QRY_CURRENCY_CODE") != null && !this.request.getParam("QRY_CURRENCY_CODE").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("pol.CURRENCY_CODE like :CURRENCY_CODE");
      this.queryParams.put("CURRENCY_CODE",(String)this.request.getParam("QRY_CURRENCY_CODE").toUpperCase());
    }
    if (this.request.getParam("QRY_NOTES") != null && !this.request.getParam("QRY_NOTES").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("pol.NOTES like :NOTES");
      this.queryParams.put("NOTES",(String)this.request.getParam("QRY_NOTES"));
    }
    if (this.request.getParam("QRY_DATE_REQUESTED") != null && !this.request.getParam("QRY_DATE_REQUESTED").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("pol.DATE_REQUESTED like :DATE_REQUESTED");
      this.queryParams.put("DATE_REQUESTED",(String)this.request.getParam("QRY_DATE_REQUESTED"));
    }
    if (this.request.getParam("QRY_DATE_PROMISED") != null && !this.request.getParam("QRY_DATE_PROMISED").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("pol.DATE_PROMISED like :DATE_PROMISED");
      this.queryParams.put("DATE_PROMISED",(String)this.request.getParam("QRY_DATE_PROMISED"));
    }
    if (this.request.getParam("QRY_DATE_DELIVERED") != null && !this.request.getParam("QRY_DATE_DELIVERED").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("pol.DATE_DELIVERED like :DATE_DELIVERED");
      this.queryParams.put("DATE_DELIVERED",(String)this.request.getParam("QRY_DATE_DELIVERED"));
    }
}

public void doQuery() throws Exception {
    this.prepareQueryContext();
    this.preQuery();
    this.queryWhere.insert(0, (this.queryWhere.length()>0)?" where ":"");
    String sql = "select "+ 
               " pol.ID"+
               " ,pol.PORDER_ID"+
               " ,pol.LINE_NR"+
               " ,pol.PRODUCT_ID"+
               " ,pol.QTY"+
               " ,pol.UOM_CODE"+
               " ,pol.PRICE"+
               " ,pol.NET_AMOUNT"+
               " ,pol.CURRENCY_CODE"+
               " ,pol.NOTES"+
               " ,pol.DATE_REQUESTED"+
               " ,pol.DATE_PROMISED"+
               " ,pol.DATE_DELIVERED"+
               " ,pbo_product.get_name_by_id(pol.product_id,'N') PRODUCT_NAME"+
           " from PURCHASE_ORDER_LINE pol "+this.queryWhere.toString()+" "+this.queryOrderBy;
    this.writeResultDoQuery(sql);
} 


public void doExport() throws Exception {
    this.prepareQueryContext();
    this.preQuery();
    this.queryWhere.insert(0, (this.queryWhere.length()>0)?" where ":"");
    String sql = "select "+ 
               " pol.ID"+
               " ,pol.PORDER_ID"+
               " ,pol.LINE_NR"+
               " ,pol.PRODUCT_ID"+
               ",pbo_product.get_name_by_id(pol.product_id,'N') PRODUCT_NAME"+
               " ,pol.QTY"+
               " ,pol.UOM_CODE"+
               " ,pol.PRICE"+
               " ,pol.CURRENCY_CODE"+
               " ,pol.NET_AMOUNT"+
               " ,pol.NOTES"+
               " ,pol.DATE_REQUESTED"+
               " ,pol.DATE_PROMISED"+
               " ,pol.DATE_DELIVERED"+
           " from PURCHASE_ORDER_LINE pol "+this.queryWhere.toString()+" "+this.queryOrderBy;
    this.writeResultDoExport(sql);
}

 	public void fetchRecord()  throws Exception  {
     this.populateRecordPkFromRequest();
     this.findByPk();
       this.writeResultFetchRecord();	 
   }

public void doInsert()  throws Exception {
  this.populateRecordFromRequest(); 
  this.populateRecordWithClientSpecific();
    String sql = "insert into PURCHASE_ORDER_LINE("+
               "  ID"+
               " ,PORDER_ID"+
               " ,LINE_NR"+
               " ,PRODUCT_ID"+
               " ,QTY"+
               " ,UOM_CODE"+
               " ,PRICE"+
               " ,NET_AMOUNT"+
               " ,CURRENCY_CODE"+
               " ,NOTES"+
               " ,DATE_REQUESTED"+
               " ,DATE_PROMISED"+
               " ,DATE_DELIVERED"+
           " ) values ( "+
               "  :ID"+
               " ,:PORDER_ID"+
               " ,:LINE_NR"+
               " ,:PRODUCT_ID"+
               " ,:QTY"+
               " ,:UOM_CODE"+
               " ,:PRICE"+
               " ,:NET_AMOUNT"+
               " ,:CURRENCY_CODE"+
               " ,:NOTES"+
               " ,:DATE_REQUESTED"+
               " ,:DATE_PROMISED"+
               " ,:DATE_DELIVERED"+
    ")";
    dbm.executeStatement(sql, this.record);
    this.populateRecordPkFromRecord();
    this.findByPk();
    this.writeResultDoInsert();
}

public void doUpdate() throws Exception {
    this.populateRecordFromRequest();
    this.populateRecordWithClientSpecific();
    String sql = "update PURCHASE_ORDER_LINE set "+
               "  ID=:ID"+
               " ,PORDER_ID=:PORDER_ID"+
               " ,LINE_NR=:LINE_NR"+
               " ,PRODUCT_ID=:PRODUCT_ID"+
               " ,QTY=:QTY"+
               " ,UOM_CODE=:UOM_CODE"+
               " ,PRICE=:PRICE"+
               " ,NET_AMOUNT=:NET_AMOUNT"+
               " ,CURRENCY_CODE=:CURRENCY_CODE"+
               " ,NOTES=:NOTES"+
               " ,DATE_REQUESTED=:DATE_REQUESTED"+
               " ,DATE_PROMISED=:DATE_PROMISED"+
               " ,DATE_DELIVERED=:DATE_DELIVERED"+
   " where "+
     "      ID= :ID"+
   "";
    dbm.executeStatement(sql, this.record);
    this.populateRecordPkFromRecord();
    this.findByPk();
    this.writeResultDoUpdate();	
}

public void doDelete() throws Exception {
    this.populateRecordPkFromRequest();
   String sql = "delete from PURCHASE_ORDER_LINE where "+
     "      ID= :ID"+
    "";
    dbm.executeStatement(sql, this.recordPk);
   this.writeResultDoDelete();
}

public void initNewRecord() throws Exception {
    this.populateRecordFromRequest();
    this.record.put("_p_record_status", "insert");
    this.writeResultInitNewRecord();
}

private void findByPk()  throws Exception {
    String sql = "select "+ 
               " pol.ID"+
               " ,pol.PORDER_ID"+
               " ,pol.LINE_NR"+
               " ,pol.PRODUCT_ID"+
               " ,pol.QTY"+
               " ,pol.UOM_CODE"+
               " ,pol.PRICE"+
               " ,pol.NET_AMOUNT"+
               " ,pol.CURRENCY_CODE"+
               " ,pol.NOTES"+
               " ,pol.DATE_REQUESTED"+
               " ,pol.DATE_PROMISED"+
               " ,pol.DATE_DELIVERED"+
                ",pbo_product.get_name_by_id(pol.product_id,'N') PRODUCT_NAME"+
           " from PURCHASE_ORDER_LINE pol"+
        " where "+
     "      pol.ID= :ID"+ 
          "";
    this.record = dbm.executeQuery(sql, this.recordPk).get(0); 
} 


public void doCustomAction(String pName)  throws Exception {
    this.populateRecordFromRequest();
}


	private void  _initFields() {
	  this.fields = new HashMap<String, FieldDef>();
	  this.fields.put("ID", new FieldDef("NUMBER"));
	  this.fields.put("PORDER_ID", new FieldDef("NUMBER"));
	  this.fields.put("LINE_NR", new FieldDef("NUMBER"));
	  this.fields.put("PRODUCT_ID", new FieldDef("NUMBER"));
	  this.fields.put("QTY", new FieldDef("NUMBER"));
	  this.fields.put("UOM_CODE", new FieldDef("STRING"));
	  this.fields.get("UOM_CODE").setCaseRestriction("Upper");
	  this.fields.put("PRICE", new FieldDef("NUMBER"));
	  this.fields.put("NET_AMOUNT", new FieldDef("NUMBER"));
	  this.fields.put("CURRENCY_CODE", new FieldDef("STRING"));
	  this.fields.get("CURRENCY_CODE").setCaseRestriction("Upper");
	  this.fields.put("NOTES", new FieldDef("STRING"));
	  this.fields.put("DATE_REQUESTED", new FieldDef("DATE"));
	  this.fields.put("DATE_PROMISED", new FieldDef("DATE"));
	  this.fields.put("DATE_DELIVERED", new FieldDef("DATE"));
	  this.fields.put("CREATEDON", new FieldDef("DATE"));
	  this.fields.get("CREATEDON").setInDS(false);
	  this.fields.put("CREATEDBY", new FieldDef("STRING"));
	  this.fields.get("CREATEDBY").setInDS(false);
	  this.fields.put("MODIFIEDON", new FieldDef("DATE"));
	  this.fields.get("MODIFIEDON").setInDS(false);
	  this.fields.put("MODIFIEDBY", new FieldDef("STRING"));
	  this.fields.get("MODIFIEDBY").setInDS(false);
	  this.fields.put("PRODUCT_NAME", new FieldDef("STRING"));
	  String[] _pkFields = {"ID"};
	  this.pkFields = _pkFields;
	  String[] _summaryFields = {};
	  this.summaryFields = _summaryFields;
	  this.queryResultSize = 20;
	}

}
